ecosystem/sf-indexer/aptos-sf-indexer/src/indexer/substream_processor.rs

```diff
@@ -0,0 +1,219 @@
+    RunQueryDsl,
+};
+use schema::indexer_states::{self, dsl};
+use std::fmt::Debug;
+
+/// The `SubstreamProcessor` processes the output from a substream specific to the server instance
+#[async_trait]
+pub trait SubstreamProcessor: Send + Sync + Debug {
+    /// name of the substream module, for status logging
+    /// This will get stored in the database for each (substream_module, block_height) pair
+    fn substream_module_name(&self) -> &'static str;
+
+    /// Accepts a transaction, and processes it. This method will be called from `process_substream_with_status`
+    /// In case a transaction cannot be processed, returns an error: the processor will mark it as failed in the database,
+    /// and it will be retried next time the indexer is started.
+    async fn process_substream(
+        &self,
+        stream_data: BlockScopedData,
+        block_height: u64,
+    ) -> Result<ProcessingResult, BlockProcessingError>;
+
+    /// Gets a reference to the connection pool
+    /// This is used by the `get_conn()` helper below
+    fn connection_pool(&self) -> &PgDbPool;
+
+    //* Below are helper methods that don't need to be implemented *//
+
+    /// Gets the connection.
+    /// If it was unable to do so (default timeout: 30s), it will keep retrying until it can.
+    /// It's a static method because we need the connection before the processor is initialized
+    fn get_conn(pool: &PgDbPool) -> PgPoolConnection {
+        loop {
+            match pool.get() {
+                Ok(conn) => {
+                    GOT_CONNECTION.inc();
+                    return conn;
+                }
+                Err(err) => {
+                    UNABLE_TO_GET_CONNECTION.inc();
+                    aptos_logger::error!(
+                        "Could not get DB connection from pool, will retry in {:?}. Err: {:?}",
+                        pool.connection_timeout(),
+                        err
+                    );
+                }
+            };
+        }
+    }
+
+    /// This is a helper method, tying together the other helper methods to allow tracking status in the DB
+    async fn process_substream_with_status(
+        &mut self,
+        current_substream_name: String,
+        stream_data: BlockScopedData,
+        block_height: u64,
+    ) -> Result<ProcessingResult, BlockProcessingError> {
+        if current_substream_name != self.substream_module_name() {
+            panic!("Wrong processor detected: this processor can only process module {},  module {} detected.", self.substream_module_name(), current_substream_name);
+        }
+        PROCESSOR_INVOCATIONS
+            .with_label_values(&[self.substream_module_name()])
+            .inc();
+
+        aptos_logger::debug!("Marking block started {}", block_height);
+        self.mark_block_started(block_height);
+        aptos_logger::debug!("Starting to process stream for block {}", block_height);
+        let res = self.process_substream(stream_data, block_height).await;
```

Comment:
do we not want to wrap this in a transaction?

Reply:
What do you mean?

Reply:
how are we ensuring that, per substream, all queries run within it are wrapped atomically? i.e within a transaction

Reply:
Inside of block_output_processor.insert_block